src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,218 @@
+    }
+
+    private Node<T> hasPrevious(Node<T> node) {
+        return node.previous != null ? node.previous : null;
+    }
+
+    private Node<T> hasNext(Node<T> node) {
+        return node.next != null ? node.next : null;
+    }
+
+    private void checkIndex(int index, boolean forAdd) {
+        if (index < 0 || index > size || (!forAdd && index == size)) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+    }
+
+    private boolean createTailIfDoesNotExist(T value) {
+        if (size == 1) {
+            tail = new Node<>(head, value, null);
+            head.next = tail;
+            size++;
+            return true;
+        }
         return false;
     }
+
+    private boolean createHeadIfDoesNotExist(T value) {
+        if (head == null) {
+            head = new Node<>(null, value, null);
+            size++;
+            return true;
+        }
+        return false;
+    }
+
+    private void updateTail(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        tail.next = newNode;
+        tail = newNode;
+        size++;
+    }
+
+    private void insertNodeAtIndex(T value, int index) {
+        Node<T> foundedNode = foundedNode(index);
+        Node<T> newNode = new Node<>(null, value, null);
+        updateRelationship(foundedNode, newNode);
+        size++;
+    }
+
+    private void updateRelationship(Node<T> foundedNode, Node<T> newNode) {
+        if (foundedNode == null) {
+            updateTail(newNode.value);
+            size--;
+        } else if (foundedNode.equals(head)) {
+            updateHeadWithAddNode(newNode);
+        } else if (foundedNode.equals(tail)) {
+            updateTailWithAddNode(tail, newNode);
+        } else {
+            updateTailWithAddNode(foundedNode, newNode);
+        }
+    }
+
+    private void updateTailWithAddNode(Node<T> tail, Node<T> newNode) {
+        tail.previous.next = newNode;
+        newNode.previous = tail.previous;
+        tail.previous = newNode;
+        newNode.next = tail;
+    }
+
+    private void updateHeadWithAddNode(Node<T> newNode) {
+        head.previous = newNode;
+        newNode.next = head;
+        head = newNode;
+    }
+
+    private void unlink(Node<T> foundedNode) {
+        if (foundedNode.equals(head)) {
+            updateHeadWithDeleteNode();
+        } else if (foundedNode.equals(tail)) {
+            updateTailWithDeleteNode();
+        } else {
+            updateChainOfNodeWithDeleteNode(foundedNode);
+        }
+    }
+
+    private void updateChainOfNodeWithDeleteNode(Node<T> foundedNode) {
+        foundedNode.previous.next = foundedNode.next;
+        foundedNode.next.previous = foundedNode.previous;
+        size--;
+    }
+
+    private void updateTailWithDeleteNode() {
+        tail.previous.next = null;
+        tail = tail.previous;
+        size--;
+    }
+
+    private void updateHeadWithDeleteNode() {
+        head = head.next;
+        size--;
+    }
```

Comment:
I personally like to change references in the method itself not in a separate one, this looks more clear to me, if you create the head and tail in the constructor, you won’t have to do so many checks and replacing references will fit in a couple of lines

Reply:
Agree, but this implementation without creating head and tail in the constructor